### PR TITLE
Move ownership of feed handlers from BmNode to BmCluster.

### PR DESCRIPTION
**Pull request by toregge**
@geirst : please review

### CODE REVIEW

File: searchcore/src/vespa/searchcore/bmcluster/storage_api_rpc_bm_feed_handler.cpp [hunk 1]
```diff
@@ -83,10 +83,20 @@ StorageApiRpcBmFeedHandler::~StorageApiRpcBmFeedHandler() = default;
 void
 StorageApiRpcBmFeedHandler::send_rpc(std::shared_ptr<storage::api::StorageCommand> cmd, PendingTracker& pending_tracker)
 {
-    cmd->setSourceIndex(0);
-    cmd->setAddress(*_storage_address);
-    _message_dispatcher->retain(cmd->getMsgId(), pending_tracker);
-    _rpc_client->send_rpc_v1_request(std::move(cmd));
+    auto bucket = cmd->getBucket();
+    if (_distributor) {
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/bmcluster/storage_api_rpc_bm_feed_handler.cpp:87]: Consider if the handler classes should inherit a common abstract class to share some member variables and code that is currently duplicated around. The same pattern is used in _StorageApiChainBmFeedHandler::send_msg_ and _StorageApiMessageBusBmFeedHandler::send_msg_.

**Review by geirst — APPROVED** [review]
👍

File: searchcore/src/vespa/searchcore/bmcluster/bm_message_bus_routes.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#pragma once
+
+#include "bm_storage_message_addresses.h"
+#include <vespa/messagebus/routing/route.h>
+
+namespace search::bmcluster {
+
+/*
+ * Class containing the message bus routes for a set of nodes at
+ * the given layer (service layer or ditributor).
```
[the remark this answers is not in the document]
Comment (toregge) [reply searchcore/src/vespa/searchcore/bmcluster/bm_message_bus_routes.h]: Fixed.

File: searchcore/src/vespa/searchcore/bmcluster/bm_storage_message_addresses.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+#include <memory>
+#include <vector>
+
+namespace storage::api { class StorageMessageAddress; }
+
+namespace search::bmcluster {
+
+/*
+ * Class representing the storage message addresses for a set of nodes at
+ * the given layer (service layer or ditributor).
```
[the remark this answers is not in the document]
Comment (toregge) [reply searchcore/src/vespa/searchcore/bmcluster/bm_storage_message_addresses.h]: Fixed.

File: searchcore/src/vespa/searchcore/bmcluster/storage_api_rpc_bm_feed_handler.cpp (hunk 1 above)
Comment (toregge) [reply searchcore/src/vespa/searchcore/bmcluster/storage_api_rpc_bm_feed_handler.cpp]: Common code has been moved to _StorageApiBmFeedHandlerBase_.

File: searchcore/src/vespa/searchcore/bmcluster/storage_api_rpc_bm_feed_handler.cpp [hunk 4]
```diff
@@ -83,10 +83,20 @@ StorageApiRpcBmFeedHandler::~StorageApiRpcBmFeedHandler() = default;
...
+        cmd->setSourceIndex(0);
+    } else {
+        cmd->setSourceIndex(_distribution.get_distributor_node_idx(bucket));
+    }
+    uint32_t node_idx = _distributor ? _distribution.get_distributor_node_idx(bucket) : _distribution.get_service_layer_node_idx(bucket);
```
[the remark this answers is not in the document]
Comment (toregge) [reply searchcore/src/vespa/searchcore/bmcluster/storage_api_rpc_bm_feed_handler.cpp]: Common code has been moved to new base class.

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**